meta/autoid/autoid.go

```diff
@@ -372,6 +439,96 @@ func CalcNeededBatchSize(base, n, increment, offset int64, isUnsigned bool) int6
 	return nr - base
 }
 
+// CalcSequenceBatchSize calculate the next sequence batch size.
+func CalcSequenceBatchSize(base, size, increment, offset, MIN, MAX int64) (int64, error) {
+	// The sequence is positive growth.
+	if increment > 0 {
+		if increment == 1 {
+			// Sequence is already allocated to the end.
+			if base >= MAX {
+				return 0, ErrAutoincReadFailed
+			}
+			// The rest of sequence < cache size, return the rest.
+			if MAX-base < size {
+				return MAX - base, nil
+			}
+			// The rest of sequence is adequate.
+			return size, nil
+		}
+		nr, ok := SeekToFirstSequenceValue(base, increment, offset, MIN, MAX)
+		if !ok {
+			return 0, ErrAutoincReadFailed
+		}
+		// The rest of sequence < cache size, return the rest.
+		if MAX-nr < (size-1)*increment {
+			return MAX - base, nil
+		}
+		return (nr - base) + (size-1)*increment, nil
+	}
+	// The sequence is negative growth.
+	if increment == -1 {
+		if base <= MIN {
+			return 0, ErrAutoincReadFailed
+		}
+		if base-MIN < size {
+			return base - MIN, nil
+		}
+		return size, nil
+	}
+	nr, ok := SeekToFirstSequenceValue(base, increment, offset, MIN, MAX)
+	if !ok {
+		return 0, ErrAutoincReadFailed
+	}
+	// The rest of sequence < cache size, return the rest.
+	if nr-MIN < (size-1)*(-increment) {
+		return base - MIN, nil
+	}
+	return (base - nr) + (size-1)*(-increment), nil
+}
+
+// SeekToFirstSequenceValue seeks to the next valid value (must be in range of [MIN, MAX]),
+// the bool indicates whether the first value is got.
+func SeekToFirstSequenceValue(base, increment, offset, MIN, MAX int64) (int64, bool) {
+	if increment > 0 {
+		// Sequence is already allocated to the end.
+		if base >= MAX {
+			return 0, false
+		}
+		// The formula will overflow cause (base + increment) > MAX (May be MaxInt64).
+		if MAX-base < increment {
+			// Enum the possible first value.
+			for i := base + 1; i <= MAX; i++ {
+				if (i-offset)%increment == 0 {
+					return i, true
+				}
+			}
+			return 0, false
+		}
+		// Get the first value with formula.
+		nr := (base + increment - offset) / increment
+		nr = nr*increment + offset
+		return nr, true
```

Comment:
It seems this enumeration is not necessary because we can first check:
```go
if base - offset > MAX - increment {
    return 0, false
}
```
and calcualte `nr` with 
```go
nr := (base - offset + increment) 
```
instead of 
```go
nr := (base + increment - offset)
```
Thus, there is no need to worry about `base + increment` overflowing.

Reply:
I find a point here, my thought and yours are all with shortcomings.
Since offset can be negative, base - offset may be huge than Int64.
eg:
Base =  MaxInt64-4,    Increment = 3,   Max = MaxInt64 
there are must be a value at least!
when Offset = MinInt64

Base-offset > MaxInt64 - Increment 爆了...

I've tried to transfer to Uint64 to compute the formula.

Reply:
I've used bigInt to compute the formula logic, PTAL